schema_generator/assessment.py

```diff
@@ -28,29 +50,33 @@ def generate_schema(framework_slug, question_set, manifest_name):
         elif "passIfIn" in question.get("assessment", {}):
             assessed_questions.append(question)
 
-    discretionary_properties = {
-        question.id: _enum_for_question(question)
-        for question in assessed_questions if question["assessment"].get("discretionary")
-    }
-    baseline_properties = {
-        question.id: _enum_for_question(question)
-        for question in assessed_questions if not question["assessment"].get("discretionary")
-    }
+    discretionary_questions = tuple(q for q in assessed_questions if q["assessment"].get("discretionary"))
+    baseline_questions = tuple(q for q in assessed_questions if not q["assessment"].get("discretionary"))
 
-    return {
+    # we add all assessed_questions as "required" by default, not because it's the assessment schema's job to
+    # enforce that requirement, but it's a stronger guarantee we keep the content, the tests for this function
+    # and the custom validation in the supplier frontend in agreement with each other.
+
+    generated_schema = {
         "$schema": "http://json-schema.org/draft-07/schema#",  # hardcoded to draft 7 because jsonschema > 3 supports it
         "title": "{} Declaration Assessment Schema (Definite Pass Schema)".format(framework_slug),
         "type": "object",
         "allOf": [
             {"$ref": "#/definitions/baseline"},
-            {"properties": discretionary_properties},
+            {"properties": {q.id: _enum_for_question(q) for q in discretionary_questions}},
         ],
+        "required": sorted(q.id for q in discretionary_questions if q["assessment"].get("required", True)),
         "definitions": {
             "baseline": {
                 "$schema": "http://json-schema.org/draft-04/schema#",
                 "title": "{} Declaration Assessment Schema (Baseline Schema)".format(framework_slug),
                 "type": "object",
-                "properties": baseline_properties,
+                "allOf": [
+                    {"properties": {q.id: _enum_for_question(q) for q in baseline_questions}},
+                ],
+                "required": sorted(q.id for q in baseline_questions if q["assessment"].get("required", True)),
```

Comment:
Have you managed to get this new schema to work with the `export-framework-results-reasons` script? When I try, the errors for the new `required` fields are output by the validator, but they have no `path` so the script (in its current form, at least) throws an error: https://github.com/alphagov/digitalmarketplace-scripts/blob/master/dmscripts/export_framework_results_reasons.py#L22

Reply:
Ugh I had a sneaking suspicion about that script...

I'll take a look - it will probably need some attention.

Reply:
Is this error happening for you when the `required` constraint is being violated? Finding this hard to reproduce.

Reply:
Handled in https://github.com/alphagov/digitalmarketplace-scripts/pull/465